File: engine/swiftui/Sources/MilanoSDK/Engine/MilanoVocabulary.swift
import Foundation

/// A parsed, validated vocabulary artifact: the consumer's component types,
/// events, and global custom actions, per the vocabulary schema spec.
struct MilanoVocabulary: Equatable, Sendable {
    struct Component: Equatable, Sendable {
        /// Property name to type.
        let properties: [String: MilanoType]
        /// Event name to payload type; `nil` payload means a payload-less event.
        let events: [String: MilanoType?]
        /// Whether nodes of this type accept `children`.
        let children: Bool
        /// When true, undeclared properties are a SchemaViolation instead of
        /// ignored-and-reported.
        let strict: Bool
    }

    struct Action: Equatable, Sendable {
        /// Parameter name to type.
        let parameters: [String: MilanoType]
        /// The success completion's value type; `nil` means completions
        /// carry no data (vocabulary schema spec, completion results).
        let result: MilanoType?
    }

    /// The contract version the artifact targets (major, minor).
    let contractMajor: Int
    let contractMinor: Int
    let name: String
    /// Consumer-owned; surfaced in observability, never interpreted.
    let version: String
    let components: [String: Component]
    let actions: [String: Action]
}

extension MilanoVocabulary {
    /// Parses and validates a vocabulary artifact from JSON bytes.
    /// Throws `MilanoEngineError.invalidVocabulary` on any rule violation.
    init(artifactJSON data: Data) throws {
        let raw: Any
        do {
            raw = try JSONSerialization.jsonObject(with: data)
        } catch {
            throw MilanoEngineError.invalidVocabulary(rule: "json", detail: "not well-formed JSON")
        }
        guard let rootJSON = MilanoValue(json: raw), case .record(let root) = rootJSON else {
            throw MilanoEngineError.invalidVocabulary(rule: "structure", detail: "artifact is not an object")
        }

        // milano: contract version "major.minor"
        guard case .string(let milano)? = root["milano"] else {
            throw MilanoEngineError.invalidVocabulary(rule: "milano", detail: "missing contract version")
        }
        let versionParts = milano.split(separator: ".", omittingEmptySubsequences: false)
        guard versionParts.count == 3,
            let major = Int(versionParts[0]), let minor = Int(versionParts[1]),
            let patch = Int(versionParts[2]),
            major >= 0, minor >= 0, patch >= 0
        else {
            throw MilanoEngineError.invalidVocabulary(
                rule: "milano", detail: "expected major.minor.patch, found \(milano)")
        }
        // Same versioning rule as documents: an artifact targeting an
        // unsupported contract major fails fast at engine creation.
        guard MilanoGate.supportedMajors.contains(major) else {
            throw MilanoEngineError.invalidVocabulary(
                rule: "milano-version",
                detail: "unsupported contract major \(major); supported: \(MilanoGate.supportedMajors)")
        }

        guard case .string(let name)? = root["name"], MilanoIdentifier.isValid(name) else {
            throw MilanoEngineError.invalidVocabulary(rule: "name", detail: "missing or invalid identifier")
        }
        guard case .string(let vocabularyVersion)? = root["version"],
            parseSemver(vocabularyVersion) != nil
        else {
            throw MilanoEngineError.invalidVocabulary(
                rule: "version", detail: "vocabulary version must be major.minor.patch")
        }

        guard case .record(let componentsJSON)? = root["components"] else {
            throw MilanoEngineError.invalidVocabulary(rule: "components", detail: "missing components")
        }
        var components: [String: Component] = [:]
        for (typeName, declaration) in componentsJSON {
            guard MilanoIdentifier.isValid(typeName) else {
                throw MilanoEngineError.invalidVocabulary(rule: "component-name", detail: typeName)
            }
            components[typeName] = try Self.component(from: declaration, at: typeName)
        }

        var actions: [String: Action] = [:]
        if let actionsEntry = root["actions"] {
            guard case .record(let actionsJSON) = actionsEntry else {
                throw MilanoEngineError.invalidVocabulary(rule: "actions", detail: "actions is not an object")
            }
            for (actionName, declaration) in actionsJSON {
                guard MilanoIdentifier.isValid(actionName) else {
                    throw MilanoEngineError.invalidVocabulary(rule: "action-name", detail: actionName)
                }
                actions[actionName] = try Self.action(from: declaration, at: actionName)
            }
        }

        self.init(
            contractMajor: major, contractMinor: minor,
            name: name, version: vocabularyVersion,
            components: components, actions: actions)
    }

    /// Parses one custom action declaration; shared with document-local
    /// declarations, which use the same format (document model spec).
    static func action(from declaration: MilanoValue, at path: String) throws -> Action {
        guard case .record(let object) = declaration else {
            throw MilanoEngineError.invalidVocabulary(rule: "action", detail: "\(path) is not an object")
        }
        var parameters: [String: MilanoType] = [:]
        if let parametersEntry = object["parameters"] {
            guard case .record(let parametersJSON) = parametersEntry else {
                throw MilanoEngineError.invalidVocabulary(rule: "action-parameters", detail: path)
            }
            for (parameterName, descriptor) in parametersJSON {
                guard MilanoIdentifier.isValid(parameterName),
                    let type = MilanoType(descriptor: descriptor)
                else {
                    throw MilanoEngineError.invalidVocabulary(
                        rule: "action-parameter", detail: "\(path).\(parameterName)")
                }
                parameters[parameterName] = type
            }
        }
        var result: MilanoType?
        if let resultEntry = object["result"] {
            guard let type = MilanoType(descriptor: resultEntry) else {
                throw MilanoEngineError.invalidVocabulary(
                    rule: "action-result", detail: path)
            }
            result = type
        }
        return Action(parameters: parameters, result: result)
    }

    private static func component(from declaration: MilanoValue, at path: String) throws -> Component {
        guard case .record(let object) = declaration else {
            throw MilanoEngineError.invalidVocabulary(rule: "component", detail: "\(path) is not an object")
        }

        var properties: [String: MilanoType] = [:]
        if let propertiesEntry = object["properties"] {
            guard case .record(let propertiesJSON) = propertiesEntry else {
                throw MilanoEngineError.invalidVocabulary(rule: "component-properties", detail: path)
            }
            for (propertyName, descriptor) in propertiesJSON {
                guard MilanoIdentifier.isValid(propertyName),
                    let type = MilanoType(descriptor: descriptor)
                else {
                    throw MilanoEngineError.invalidVocabulary(
                        rule: "component-property", detail: "\(path).\(propertyName)")
                }
                properties[propertyName] = type
            }
        }

        var events: [String: MilanoType?] = [:]
        if let eventsEntry = object["events"] {
            guard case .record(let eventsJSON) = eventsEntry else {
                throw MilanoEngineError.invalidVocabulary(rule: "component-events", detail: path)
            }
            for (eventName, descriptor) in eventsJSON {
                guard MilanoIdentifier.isValid(eventName) else {
                    throw MilanoEngineError.invalidVocabulary(
                        rule: "component-event", detail: "\(path).\(eventName)")
                }
                if descriptor == .null {
                    events[eventName] = MilanoType?.none
                } else if let payloadType = MilanoType(descriptor: descriptor) {
                    events[eventName] = payloadType
                } else {
                    throw MilanoEngineError.invalidVocabulary(
                        rule: "component-event", detail: "\(path).\(eventName)")
                }
            }
        }

        let children: Bool
        switch object["children"] {
        case nil: children = false
        case .bool(let flag): children = flag
        default:
            throw MilanoEngineError.invalidVocabulary(rule: "component-children", detail: path)
        }

        let strict: Bool
        switch object["strict"] {
        case nil: strict = false
        case .bool(let flag): strict = flag
        default:
            throw MilanoEngineError.invalidVocabulary(rule: "component-strict", detail: path)
        }

        return Component(properties: properties, events: events, children: children, strict: strict)
    }
}

File: engine/compose/src/commonMain/kotlin/dev/getmilano/engine/MilanoVocabulary.kt
package dev.getmilano

import kotlinx.serialization.json.Json

/**
 * A parsed, validated vocabulary artifact: the consumer's component types,
 * events, and global custom actions, per the vocabulary schema spec.
 */
internal data class MilanoVocabulary(
    /** The contract version the artifact targets. */
    val contractMajor: Int,
    val contractMinor: Int,
    val name: String,
    /** Consumer-owned; surfaced in observability, never interpreted. */
    val version: String,
    val components: Map<String, Component>,
    val actions: Map<String, Action>,
) {
    data class Component(
        /** Property name to type. */
        val properties: Map<String, MilanoType>,
        /**
         * Event name to payload type; a null payload means a payload-less
         * event. Presence in the map is what "declared" means.
         */
        val events: Map<String, MilanoType?>,
        /** Whether nodes of this type accept children. */
        val children: Boolean,
        /**
         * When true, undeclared properties are a SchemaViolation instead of
         * ignored-and-reported.
         */
        val strict: Boolean,
    )

    data class Action(
        /** Parameter name to type. */
        val parameters: Map<String, MilanoType>,
        /**
         * The success completion's value type; null means completions carry
         * no data (vocabulary schema spec, completion results).
         */
        val result: MilanoType? = null,
    )

    companion object {
        /**
         * Parses and validates a vocabulary artifact from JSON text.
         * Throws [MilanoEngineException.InvalidVocabulary] on any rule violation.
         */
        fun parse(artifactJson: String): MilanoVocabulary {
            val root =
                try {
                    MilanoValue.fromJson(Json.parseToJsonElement(artifactJson))
                } catch (_: Exception) {
                    throw MilanoEngineException.InvalidVocabulary("json", "not well-formed JSON")
                }
            val rootRecord =
                (root as? MilanoValue.RecordValue)?.values
                    ?: throw MilanoEngineException.InvalidVocabulary("structure", "artifact is not an object")

            val milano =
                (rootRecord["milano"] as? MilanoValue.StringValue)?.value
                    ?: throw MilanoEngineException.InvalidVocabulary("milano", "missing contract version")
            val versionParts = milano.split(".")
            val major = versionParts.getOrNull(0)?.toIntOrNull()
            val minor = versionParts.getOrNull(1)?.toIntOrNull()
            val patch = versionParts.getOrNull(2)?.toIntOrNull()
            if (versionParts.size != 3 || major == null || minor == null || patch == null ||
                major < 0 || minor < 0 || patch < 0
            ) {
                throw MilanoEngineException.InvalidVocabulary("milano", "expected major.minor.patch, found $milano")
            }
            // Same versioning rule as documents: an artifact targeting an
            // unsupported contract major fails fast at engine creation.
            if (major !in MilanoGate.SUPPORTED_MAJORS) {
                throw MilanoEngineException.InvalidVocabulary(
                    "milano-version",
                    "unsupported contract major $major; supported: ${MilanoGate.SUPPORTED_MAJORS}",
                )
            }

            val name =
                (rootRecord["name"] as? MilanoValue.StringValue)
                    ?.value
                    ?.takeIf { MilanoIdentifier.isValid(it) }
                    ?: throw MilanoEngineException.InvalidVocabulary("name", "missing or invalid identifier")
            val vocabularyVersion =
                (rootRecord["version"] as? MilanoValue.StringValue)
                    ?.value
                    ?.takeIf { parseSemver(it) != null }
                    ?: throw MilanoEngineException.InvalidVocabulary(
                        "version",
                        "vocabulary version must be major.minor.patch",
                    )

            val componentsJson =
                (rootRecord["components"] as? MilanoValue.RecordValue)?.values
                    ?: throw MilanoEngineException.InvalidVocabulary("components", "missing components")
            val components = LinkedHashMap<String, Component>(componentsJson.size)
            for ((typeName, declaration) in componentsJson) {
                if (!MilanoIdentifier.isValid(typeName)) {
                    throw MilanoEngineException.InvalidVocabulary("component-name", typeName)
                }
                components[typeName] = component(declaration, typeName)
            }

            val actions = LinkedHashMap<String, Action>()
            when (val actionsEntry = rootRecord["actions"]) {
                null -> {}

                is MilanoValue.RecordValue -> {
                    for ((actionName, declaration) in actionsEntry.values) {
                        if (!MilanoIdentifier.isValid(actionName)) {
                            throw MilanoEngineException.InvalidVocabulary("action-name", actionName)
                        }
                        actions[actionName] = action(declaration, actionName)
                    }
                }

                else -> {
                    throw MilanoEngineException.InvalidVocabulary("actions", "actions is not an object")
                }
            }

            return MilanoVocabulary(major, minor, name, vocabularyVersion, components, actions)
        }

        /**
         * Parses one custom action declaration; shared with document-local
         * declarations, which use the same format (document model spec).
         */
        internal fun action(
            declaration: MilanoValue,
            path: String,
        ): Action {
            val record =
                (declaration as? MilanoValue.RecordValue)?.values
                    ?: throw MilanoEngineException.InvalidVocabulary("action", "$path is not an object")
            val parameters = LinkedHashMap<String, MilanoType>()
            when (val parametersEntry = record["parameters"]) {
                null -> {}

                is MilanoValue.RecordValue -> {
                    for ((parameterName, descriptor) in parametersEntry.values) {
                        val type = MilanoType.fromDescriptor(descriptor)
                        if (!MilanoIdentifier.isValid(parameterName) || type == null) {
                            throw MilanoEngineException.InvalidVocabulary("action-parameter", "$path.$parameterName")
                        }
                        parameters[parameterName] = type
                    }
                }

                else -> {
                    throw MilanoEngineException.InvalidVocabulary("action-parameters", path)
                }
            }
            var result: MilanoType? = null
            val resultEntry = record["result"]
            if (resultEntry != null) {
                result = MilanoType.fromDescriptor(resultEntry)
                    ?: throw MilanoEngineException.InvalidVocabulary("action-result", path)
            }
            return Action(parameters, result)
        }

        private fun component(
            declaration: MilanoValue,
            path: String,
        ): Component {
            val record =
                (declaration as? MilanoValue.RecordValue)?.values
                    ?: throw MilanoEngineException.InvalidVocabulary("component", "$path is not an object")

            val properties = LinkedHashMap<String, MilanoType>()
            when (val propertiesEntry = record["properties"]) {
                null -> {}

                is MilanoValue.RecordValue -> {
                    for ((propertyName, descriptor) in propertiesEntry.values) {
                        val type = MilanoType.fromDescriptor(descriptor)
                        if (!MilanoIdentifier.isValid(propertyName) || type == null) {
                            throw MilanoEngineException.InvalidVocabulary("component-property", "$path.$propertyName")
                        }
                        properties[propertyName] = type
                    }
                }

                else -> {
                    throw MilanoEngineException.InvalidVocabulary("component-properties", path)
                }
            }

            val events = LinkedHashMap<String, MilanoType?>()
            when (val eventsEntry = record["events"]) {
                null -> {}

                is MilanoValue.RecordValue -> {
                    for ((eventName, descriptor) in eventsEntry.values) {
                        if (!MilanoIdentifier.isValid(eventName)) {
                            throw MilanoEngineException.InvalidVocabulary("component-event", "$path.$eventName")
                        }
                        if (descriptor is MilanoValue.Null) {
                            events[eventName] = null
                        } else {
                            events[eventName] = MilanoType.fromDescriptor(descriptor)
                                ?: throw MilanoEngineException.InvalidVocabulary("component-event", "$path.$eventName")
                        }
                    }
                }

                else -> {
                    throw MilanoEngineException.InvalidVocabulary("component-events", path)
                }
            }

            val children =
                when (val flag = record["children"]) {
                    null -> false
                    is MilanoValue.BoolValue -> flag.value
                    else -> throw MilanoEngineException.InvalidVocabulary("component-children", path)
                }
            val strict =
                when (val flag = record["strict"]) {
                    null -> false
                    is MilanoValue.BoolValue -> flag.value
                    else -> throw MilanoEngineException.InvalidVocabulary("component-strict", path)
                }

            return Component(properties, events, children, strict)
        }
    }
}
